File: dlang/plugin-api/src/main/java/io/github/intellij/dlanguage/parser/TrailingDocCommentBinder.java
package io.github.intellij.dlanguage.parser;

import com.intellij.lang.WhitespacesAndCommentsBinder;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.psi.tree.IElementType;
import io.github.intellij.dlanguage.psi.DTokenSets;

import java.util.List;

public class TrailingDocCommentBinder implements WhitespacesAndCommentsBinder {

    public static final TrailingDocCommentBinder INSTANCE = new TrailingDocCommentBinder();

    @Override
    public int getEdgePosition(List<? extends IElementType> tokens, boolean atStreamEdge, TokenTextGetter getter) {
        // bind trailing doc comments to their declaration only if they are on the same line
        int candidate = 0;
        while (candidate < tokens.size()) {
            if (DTokenSets.WHITESPACES.contains(tokens.get(candidate)) && StringUtil.containsLineBreak(getter.get(candidate))) {
                break;
            }
            candidate++;
        }
        return candidate;
    }
}

File: dlang/psi-impl/src/main/kotlin/io/github/intellij/dlanguage/parser/TrailingDocCommentBinder.kt
package io.github.intellij.dlanguage.parser

import com.intellij.lang.WhitespacesAndCommentsBinder
import com.intellij.openapi.util.text.StringUtil
import com.intellij.psi.tree.IElementType
import io.github.intellij.dlanguage.psi.DTokenSets

class TrailingDocCommentBinder : WhitespacesAndCommentsBinder {
    override fun getEdgePosition(
        tokens: List<IElementType?>,
        atStreamEdge: Boolean,
        getter: WhitespacesAndCommentsBinder.TokenTextGetter
    ): Int {
        // bind trailing doc comments to their declaration only if they are on the same line
        var candidate = 0
        while (candidate < tokens.size) {
            if (DTokenSets.WHITESPACES.contains(tokens[candidate]) && StringUtil.containsLineBreak(
                    getter[candidate]
                )
            ) {
                break
            }
            candidate++
        }
        return candidate
    }

    companion object {
        @JvmField
        val INSTANCE: TrailingDocCommentBinder = TrailingDocCommentBinder()
    }
}
